File: t/CS/MainDemo/UI/Command/BrightnessDialog.cs
﻿// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using Leadtools;
using Leadtools.ImageProcessing;
using Leadtools.ImageProcessing.Color;
using Leadtools.Drawing;

namespace MainDemo
{
   public partial class BrightnessDialog : Form
   {
      private ImagePreviewCtrl _beforeViewer;
      private ImagePreviewCtrl _afterViewer;
      private RasterImage _originalImage;
      private RasterImage _afterImage;
      private int _internalBrightnessValue;
      private int _brightness;

      public int Brightness
      {
         set
         {
            _brightness = value;
         }
         get
         {
            return _brightness;
         }
      }

      public BrightnessDialog(RasterImage image, RasterPaintProperties paintProperties)
      {
         try
         {
            InitializeComponent();

            if (image != null)
            {
               CloneCommand clone = new CloneCommand();

               clone.Run(image);

               _originalImage = image;
               _afterImage = clone.DestinationImage;

               _beforeViewer = new ImagePreviewCtrl(_originalImage, paintProperties, _lblBefore.Location, _lblBefore.Size);
               _afterViewer = new ImagePreviewCtrl(_afterImage, paintProperties, _lblAfter.Location, _lblAfter.Size);

               Controls.Add(_beforeViewer);
               Controls.Add(_afterViewer);
               _beforeViewer.BringToFront();
               _afterViewer.BringToFront();

               _beforeViewer.PanImage += new EventHandler<PanImageEvent>(_beforeViewer_PanImage);
               _afterViewer.PanImage += new EventHandler<PanImageEvent>(_afterViewer_PanImage);
            }
            else
            {
               _tsZoomLevel.Visible = false;
            }

            _internalBrightnessValue = 0;
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      void _beforeViewer_PanImage(object sender, PanImageEvent e)
      {
         _afterViewer.OffsetImage(e.Offset);
      }

      void _afterViewer_PanImage(object sender, PanImageEvent e)
      {
         _beforeViewer.OffsetImage(e.Offset);
      }


      protected void UpdateValues()
      {
         try
         {
            RasterImage tempImage;
            CloneCommand clone = new CloneCommand();

            clone.Run(_originalImage);

            tempImage = clone.DestinationImage;

            if (DoEffect(ref tempImage))
            {
               if (_afterImage != null)
               {
                  _afterImage.Dispose();

                  _afterImage = null;
               }

               _afterImage = tempImage;

               _afterViewer.Image = _afterImage;

               _afterViewer.OffsetImage(_beforeViewer.Offset);

               _afterViewer.Invalidate();
            }
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      protected virtual bool DoEffect(ref RasterImage effectedImage)
      {
         try
         {
            ChangeIntensityCommand command = new ChangeIntensityCommand(_internalBrightnessValue);

            command.Progress += new EventHandler<RasterCommandProgressEventArgs>(command_Progress);

            if (RasterImageChangedFlags.None == command.Run(effectedImage))
            {
               return false;
            }

            //Reset progress bar
            _pbProgress.Value = 0;

            return true;
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      void command_Progress(object sender, RasterCommandProgressEventArgs e)
      {
         _pbProgress.Value = e.Percent;
      }

      private void _tsbtnNormal_Click(object sender, EventArgs e)
      {
         try
         {
            if (_beforeViewer.Image != null)
            {
               _tsbtnFit.Checked = false;
               _tsbtnNormal.Checked = true;

               _beforeViewer.FitView = false;
               _afterViewer.FitView = false;
            }
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      private void _tsbtnFit_Click(object sender, EventArgs e)
      {
         try
         {
            if (_beforeViewer.Image != null)
            {
               _tsbtnFit.Checked = true;
               _tsbtnNormal.Checked = false;

               _beforeViewer.FitView = true;
               _afterViewer.FitView = true;
            }
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      private void BrightnessDialog_Load(object sender, EventArgs e)
      {
         try
         {
            if (_beforeViewer.Image != null)
            {
               _tsbtnFit.Checked = false;
               _tsbtnNormal.Checked = true;
            }

            _internalBrightnessValue = Brightness;

            if ((_internalBrightnessValue / 10) < _numValue.Minimum)
            {
               _internalBrightnessValue = Brightness = (int)_numValue.Minimum;
            }

            if ((_internalBrightnessValue / 10) > _numValue.Maximum)
            {
               _internalBrightnessValue = Brightness = (int)_numValue.Maximum;
            }

            _numValue.Value = _internalBrightnessValue / 10;
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      private void _num_Leave(object sender, System.EventArgs e)
      {
         try
         {
            int newValue = 0;

            if (_numValue.Value < _numValue.Minimum)
            {
               _numValue.Value = _numValue.Minimum;
            }
            else if (_numValue.Value > _numValue.Maximum)
            {
               _numValue.Value = _numValue.Maximum;
            }

            newValue = (int)_numValue.Value * 10;

            if (_internalBrightnessValue != newValue)
            {
               _internalBrightnessValue = newValue;

               UpdateValues();
            }
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      void _numValue_ValueChanged(object sender, System.EventArgs e)
      {
         try
         {
            _internalBrightnessValue = (int)_numValue.Value * 10;

            UpdateValues();
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }

      private void _btnOk_Click(object sender, System.EventArgs e)
      {
         try
         {
            Brightness = _internalBrightnessValue;
         }
         catch (Exception ex)
         {
            throw ex;
         }
      }
   }
}

File: t/VB/MainDemo/UI/Command/BrightnessDialog.vb
﻿' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.ImageProcessing
Imports Leadtools.ImageProcessing.Color
Imports Leadtools.Drawing

Namespace MainDemo
   Partial Public Class BrightnessDialog : Inherits Form
      Private WithEvents _beforeViewer As ImagePreviewCtrl
      Private WithEvents _afterViewer As ImagePreviewCtrl
      Private _originalImage As RasterImage
      Private _afterImage As RasterImage
      Private _internalBrightnessValue As Integer
      Private _brightness As Integer

      Public Property Brightness() As Integer
         Set(ByVal value As Integer)
            _brightness = value
         End Set
         Get
            Return _brightness
         End Get
      End Property

      Public Sub New(ByVal image As RasterImage, ByVal paintProperties As RasterPaintProperties)
         Try
            InitializeComponent()

            If Not image Is Nothing Then
               Dim clone As CloneCommand = New CloneCommand()

               clone.Run(image)

               _originalImage = image
               _afterImage = clone.DestinationImage

               _beforeViewer = New ImagePreviewCtrl(_originalImage, paintProperties, _lblBefore.Location, _lblBefore.Size)
               _afterViewer = New ImagePreviewCtrl(_afterImage, paintProperties, _lblAfter.Location, _lblAfter.Size)

               Controls.Add(_beforeViewer)
               Controls.Add(_afterViewer)
               _beforeViewer.BringToFront()
               _afterViewer.BringToFront()
            Else
               _tsZoomLevel.Visible = False
            End If

            _internalBrightnessValue = 0
         Catch ex As Exception
            Throw ex
         End Try
      End Sub

      Protected Sub UpdateValues()
         Try
            If Not (_originalImage Is Nothing) Then
               Dim tempImage As RasterImage
               Dim clone As CloneCommand = New CloneCommand()

               clone.Run(_originalImage)

               tempImage = clone.DestinationImage

               If DoEffect(tempImage) Then
                  If Not _afterImage Is Nothing Then
                     _afterImage.Dispose()

                     _afterImage = Nothing
                  End If

                  _afterImage = tempImage

                  _afterViewer.Image = _afterImage

                  _afterViewer.OffsetImage(_beforeViewer.Offset)

                  _afterViewer.Invalidate()
               End If
            End If
         Catch ex As Exception
            Throw ex
         End Try
      End Sub

      Protected Overridable Function DoEffect(ByRef effectedImage As RasterImage) As Boolean
         Try
            Dim command As ChangeIntensityCommand = New ChangeIntensityCommand(_internalBrightnessValue)

            AddHandler command.Progress, AddressOf command_Progress

            If RasterImageChangedFlags.None = command.Run(effectedImage) Then
               Return False
            End If

            'Reset progress bar
            _pbProgress.Value = 0

            Return True
         Catch ex As Exception
            Throw ex
         End Try
      End Function

      Private Sub command_Progress(ByVal sender As Object, ByVal e As RasterCommandProgressEventArgs)
         _pbProgress.Value = e.Percent
      End Sub

      Private Sub _tsbtnNormal_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _tsbtnNormal.Click
         Try
            If Not _beforeViewer.Image Is Nothing Then
               _tsbtnFit.Checked = False
               _tsbtnNormal.Checked = True

               _beforeViewer.FitView = False
               _afterViewer.FitView = False
            End If
         Catch ex As Exception
            Throw ex
         End Try
      End Sub

      Private Sub _tsbtnFit_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _tsbtnFit.Click
         Try
            If Not _beforeViewer.Image Is Nothing Then
               _tsbtnFit.Checked = True
               _tsbtnNormal.Checked = False

               _beforeViewer.FitView = True
               _afterViewer.FitView = True
            End If
         Catch ex As Exception
            Throw ex
         End Try
      End Sub

      Private Sub BrightnessDialog_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
         Try
            If Not _beforeViewer.Image Is Nothing Then
               _tsbtnFit.Checked = False
               _tsbtnNormal.Checked = True
            End If

            _internalBrightnessValue = Brightness

            If ((_internalBrightnessValue / 10) < _numValue.Minimum) Then
               _internalBrightnessValue = CInt(_numValue.Minimum)
               Brightness = CInt(_numValue.Minimum)
            End If

            If ((_internalBrightnessValue / 10) > _numValue.Maximum) Then
               _internalBrightnessValue = CInt(_numValue.Maximum)
               Brightness = CInt(_numValue.Maximum)
            End If

            _numValue.Value = CInt(_internalBrightnessValue / 10)
         Catch ex As Exception
            Throw ex
         End Try
      End Sub

      Private Sub _num_Leave(ByVal sender As Object, ByVal e As System.EventArgs) Handles _numValue.Leave
         Try
            If _numValue.Value < _numValue.Minimum Then
               _numValue.Value = _numValue.Minimum
            ElseIf _numValue.Value > _numValue.Maximum Then
               _numValue.Value = _numValue.Maximum
            End If

            _internalBrightnessValue = CInt(_numValue.Value) * 10

            UpdateValues()
         Catch ex As Exception
            Throw ex
         End Try
      End Sub

      Private Sub _numValue_ValueChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles _numValue.ValueChanged
         Try
            _internalBrightnessValue = CInt(_numValue.Value) * 10

            UpdateValues()
         Catch ex As Exception
            Throw ex
         End Try
      End Sub

      Private Sub _btnOk_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles _btnOk.Click
         Try
            Brightness = _internalBrightnessValue
         Catch ex As Exception
            Throw ex
         End Try
      End Sub
      Private Sub _beforeViewer_PanImage(ByVal sender As Object, ByVal e As PanImageEvent) Handles _beforeViewer.PanImage
         _afterViewer.OffsetImage(e.Offset)
      End Sub
      Private Sub _afterViewer_PanImage(ByVal sender As Object, ByVal e As PanImageEvent) Handles _afterViewer.PanImage
         _beforeViewer.OffsetImage(e.Offset)
      End Sub
   End Class
End Namespace

